pyega3/pyega3.py

```diff
@@ -38,606 +33,706 @@
+            with open(self.get_fname_md5(output_file),
+                      'wb') as md5_file:  # save good md5 in aux file for future re-use
+                md5_file.write(received_file_md5.encode())
+        else:
+            os.remove(output_file)
+            raise Exception(
+                "Download process expected md5 value '{}' but got '{}'".format(check_sum, received_file_md5))
+
+    def download_file_retry(self, creds, file_id, display_file_name, file_name, file_size, check_sum, num_connections,
+                            key, output_file, genomic_range_args, max_retries, retry_wait):
+        time0 = time.time()
+        token = self.get_token(creds)
+
+        if file_name.endswith(".gpg"):
+            logging.info("GPG files are not supported")
+            return
+
+        logging.info("File Id: '{}'({} bytes).".format(file_id, file_size))
+
+        if output_file is None:
+            output_file = self.generate_output_filename(os.getcwd(), file_id, display_file_name, genomic_range_args)
+        directory = os.path.dirname(output_file)
+        if not os.path.exists(directory) and len(directory) > 0:
+            os.makedirs(directory)
+
+        hdd = psutil.disk_usage(os.getcwd())
+        logging.info("Total space : {:.2f} GiB".format(hdd.total / (2 ** 30)))
+        logging.info("Used space : {:.2f} GiB".format(hdd.used / (2 ** 30)))
+        logging.info("Free space : {:.2f} GiB".format(hdd.free / (2 ** 30)))
+
+        if self.is_genomic_range(genomic_range_args):
+            with open(output_file, 'wb') as output:
+                htsget.get(
+                    self.url_api_ticket + "/tickets/files/{}".format(file_id),
+                    output,
+                    reference_name=genomic_range_args[0], reference_md5=genomic_range_args[1],
+                    start=genomic_range_args[2], end=genomic_range_args[3],
+                    data_format=genomic_range_args[4],
+                    max_retries=sys.maxsize if max_retries < 0 else max_retries,
+                    retry_wait=retry_wait,
+                    bearer_token=token)
+            self.print_local_file_info_genomic_range('Saved to : ', output_file, genomic_range_args)
+            return
+
+        done = False
+        num_retries = 0
+        while not done:
+            try:
+                if time.time() - time0 > 1 * 60 * 60:  # token expires in 1 hour
+                    time0 = time.time()
+                    token = self.get_token(creds)
+                self.download_file(token, file_id, file_size, check_sum, num_connections, key, output_file)
+                done = True
+            except Exception as ex:
+                logging.exception(ex)
+
+                if num_retries == max_retries:
+                    if self.temporary_files_should_be_deleted:
+                        self.delete_temporary_files(self.temporary_files)
```

Comment:
Currently, the clean-up (i.e. the removal of the temporary files) only happens in the following cases:

1. The user specifically requested it on the command-line (i.e. "fetch --delete-temp-files") _AND_ there was an exception (line 495) _AND_ the program exhausted its maximum number of retry attempts (line 498). Before exiting, the program cleans up after itself (line 500).

1. The user specifically requested a clean-up on the command-line with the "clean" sub-command.

There's no clean-up performed when the user interrupts the program with Ctrl-C during a long and boring download. We could, therefore, also catch the KeyboardInterrupt exception and the SIGINT signal (these happen when the user interrupts the program with Ctrl-C) and perform a clean-up just before the program exits.

Should we do that?
Should I do that in this ticket or maybe later?